python/lsst/obs/base/cameraMapper.py

```diff
@@ -1028,36 +1029,100 @@ def _setCcdDetector(self, item, dataId, trimmed=True):
+
+        Parameters
+        ----------
+        definitions : `lsst.obs.base.FilterDefinitionCollection`
+            The filter definitions in which to search for filters.
+        idFilter : `str` or `None`
+            The filter information provided in a data ID.
+        filterLabel : `lsst.afw.image.FilterLabel` or `None`
+            The filter information provided by an exposure; may be incomplete.
+
+        Returns
+        -------
+        filters : `set` [`lsst.obs.base.FilterDefinition`]
+            The set of filters consistent with ``idFilter``
+            and ``filterLabel``.
+        """
+        # Assume none of the filter constraints actually wrong/contradictory.
+        # Then taking the intersection of all constraints will give a unique
+        # result if one exists.
+        matches = set(definitions)
+        if idFilter:
+            matches.intersection_update(definitions.findAll(idFilter))
+        if filterLabel and filterLabel.hasPhysicalLabel():
+            matches.intersection_update(definitions.findAll(filterLabel.physicalLabel))
+        if filterLabel and filterLabel.hasBandLabel():
+            matches.intersection_update(definitions.findAll(filterLabel.bandLabel))
+        return matches
+
     def _setFilter(self, mapping, item, dataId):
-        """Set the filter object in an Exposure.  If the Exposure had a FILTER
-        keyword, this was already processed during load.  But if it didn't,
-        use the filter from the registry.
+        """Set the filter information in an Exposure.
+
+        The Exposure should already have had a filter loaded, but the reader
+        (in ``afw``) had to act on incomplete information. This method
+        cross-checks the filter against the data ID and the standard list
+        of filters.
 
         Parameters
         ----------
         mapping : `lsst.obs.base.Mapping`
-            Where to get the filter from.
+            Where to get the data ID filter from.
         item : `lsst.afw.image.Exposure`
             Exposure to set the filter in.
         dataId : `dict`
             Dataset identifier.
         """
-
         if not (isinstance(item, afwImage.ExposureU) or isinstance(item, afwImage.ExposureI)
                 or isinstance(item, afwImage.ExposureF) or isinstance(item, afwImage.ExposureD)):
             return
 
-        if item.getFilter().getId() != afwImage.Filter.UNKNOWN:
-            return
-
-        actualId = mapping.need(['filter'], dataId)
-        filterName = actualId['filter']
-        if self.filters is not None and filterName in self.filters:
-            filterName = self.filters[filterName]
         try:
-            item.setFilter(afwImage.Filter(filterName))
-        except pexExcept.NotFoundError:
-            self.log.warn("Filter %s not defined.  Set to UNKNOWN." % (filterName))
+            filterDefinitions = self.getGen3Instrument()().filterDefinitions
```

Comment:
Because `getGen3Instrument()` returns an `Instrument` class, and not an object (though all `Instrument` subclasses are required to be constructible with no arguments). No idea why it was designed that way.

As for the Gen 2/3 distinction, I think `Instrument` was intended to be a bridge between the two, so it's not a violation to use it in Gen 2 code. Though I'll admit I'm calling it mostly out of laziness (there is a way to get at the filter definitions without using `Instrument`, but it would require making an API change to `CameraMapper` and all its subclasses, and `getGen3Instrument` was already there).